{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "8627d90e-243a-4213-9ce8-9a0d2e97fa32",
   "metadata": {},
   "outputs": [],
   "source": [
    "from pathlib import Path\n",
    "from typing import Optional\n",
    "from functools import reduce\n",
    "\n",
    "import typer\n",
    "import pyranges as pr\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "c074794a-f143-410a-bdc7-63014a5f0318",
   "metadata": {},
   "outputs": [],
   "source": [
    "def add_source(annotation_file: Path) -> pr.PyRanges:\n",
    "\n",
    "    prf = pr.read_gff(annotation_file)\n",
    "    samplename = pd.Series(\n",
    "        data=[annotation_file.parts[-2] for _ in range(len(prf))],\n",
    "        name=\"Source\"\n",
    "    )\n",
    "\n",
    "    return prf.insert(samplename)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "0b85acfa-7ea1-4822-b998-b7edcd9139e7",
   "metadata": {},
   "outputs": [],
   "source": [
    "def merge_ranges(pr1: pr.PyRanges, pr2: pr.PyRanges) -> pr.PyRanges:\n",
    "\n",
    "    df1 = pr1.as_df()\n",
    "    df2 = pr2.as_df()\n",
    "\n",
    "    merged_df = pd.concat([df1, df2], ignore_index=True, axis=0)\n",
    "\n",
    "    return pr.PyRanges(df=merged_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "5bb7650b-9681-431e-be4c-9a60a7a416f4",
   "metadata": {},
   "outputs": [],
   "source": [
    "from glob import glob"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "a2586d26-c399-434d-babb-e7f73a1759e2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Help on function glob in module glob:\n",
      "\n",
      "glob(pathname, *, recursive=False)\n",
      "    Return a list of paths matching a pathname pattern.\n",
      "    \n",
      "    The pattern may contain simple shell-style wildcards a la\n",
      "    fnmatch. However, unlike fnmatch, filenames starting with a\n",
      "    dot are special cases that are not matched by '*' and '?'\n",
      "    patterns.\n",
      "    \n",
      "    If recursive is true, the pattern '**' will match any files and\n",
      "    zero or more directories and subdirectories.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "help(glob)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "9bbc2a57-639a-4190-8c92-36341da40421",
   "metadata": {},
   "outputs": [],
   "source": [
    "from os import getcwd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "6ebff112-549f-456e-8f37-c909b6ea3a77",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/home/milo/workspace/cdna_brownie/notebooks'"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "getcwd()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "09684d57-1f06-4997-b82d-395e549e3095",
   "metadata": {},
   "outputs": [],
   "source": [
    "directories = glob(pathname = \"./test_data/bc10[0-9]?\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "ddf12e7d-7ead-4add-b848-66dbf9a8c0fb",
   "metadata": {},
   "outputs": [],
   "source": [
    "group_filename = \"repaired.sample.group.txt\"\n",
    "gff_filename = \"repaired.sample.gff\"\n",
    "count_filename = \"repaired.sample.abundance.txt\"\n",
    "fastq_filename = None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "8c78afab-4ff5-4f5f-ba81-69ec69680d5f",
   "metadata": {},
   "outputs": [],
   "source": [
    "directories = [Path(_) for _ in directories]\n",
    "datafiles = {\n",
    "        \"group\": group_filename,\n",
    "        \"annotation\": gff_filename,\n",
    "        \"counts\": count_filename\n",
    "    }\n",
    "\n",
    "if fastq_filename:\n",
    "    datafiles.update({\"sequence\": fastq_filename})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "1c8f1866-d944-4413-bd22-3d41c4bb344d",
   "metadata": {},
   "outputs": [],
   "source": [
    "for x in datafiles:\n",
    "        item = [_ for _ in datafiles[x].split(\",\")]\n",
    "        if len(item) == 1:\n",
    "            datafiles[x] = [a.joinpath(item[0]) for a in directories]\n",
    "        elif len(item) == len(directories):\n",
    "            datafiles[x] = [a.joinpath(b) for a, b in zip(directories, item)]\n",
    "        else:\n",
    "            raise RuntimeError(f\"The number of filenames given for {x} does not match the number of directories.\")\n",
    "\n",
    "        # Raise an issue if the given files do not exist\n",
    "        for i in datafiles[x]:\n",
    "            if not i.exists():\n",
    "                raise FileNotFoundError(f\"The {x} file {i} cannot be found\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "8479f94d-ac59-4bc9-8c01-f6ed79694b23",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[PosixPath('test_data/bc1002/repaired.sample.gff'),\n",
       " PosixPath('test_data/bc1018/repaired.sample.gff'),\n",
       " PosixPath('test_data/bc1012/repaired.sample.gff'),\n",
       " PosixPath('test_data/bc1019/repaired.sample.gff'),\n",
       " PosixPath('test_data/bc1004/repaired.sample.gff'),\n",
       " PosixPath('test_data/bc1006/repaired.sample.gff'),\n",
       " PosixPath('test_data/bc1023/repaired.sample.gff'),\n",
       " PosixPath('test_data/bc1020/repaired.sample.gff'),\n",
       " PosixPath('test_data/bc1003/repaired.sample.gff'),\n",
       " PosixPath('test_data/bc1001/repaired.sample.gff'),\n",
       " PosixPath('test_data/bc1005/repaired.sample.gff'),\n",
       " PosixPath('test_data/bc1008/repaired.sample.gff')]"
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "datafiles[\"annotation\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "85f70e7e-4366-431b-9cdc-905a57c6810c",
   "metadata": {},
   "outputs": [],
   "source": [
    "annotations = map(add_source, datafiles[\"annotation\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "45f83a75-b207-41c3-9c38-ced822cc7d0c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "PosixPath('test_data/bc1002/repaired.sample.gff')"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "datafiles[\"annotation\"][0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "d56f0e74-8fc9-4160-92a5-5e068aae36b2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'bc1002'"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Path(datafiles[\"annotation\"][0]).parts[-2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "bba795b8-6804-4416-8ad2-e079131c4d2f",
   "metadata": {},
   "outputs": [],
   "source": [
    "bc1002_annotation = add_source(datafiles[\"annotation\"][0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "744210e3-eeb6-471e-8be8-580461084878",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>bc1002</td>\n",
       "      <td>transcript</td>\n",
       "      <td>854801</td>\n",
       "      <td>859446</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PB.2</td>\n",
       "      <td>PB.2.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>bc1002</td>\n",
       "      <td>exon</td>\n",
       "      <td>854801</td>\n",
       "      <td>859446</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PB.2</td>\n",
       "      <td>PB.2.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>bc1002</td>\n",
       "      <td>transcript</td>\n",
       "      <td>2133698</td>\n",
       "      <td>2139623</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PB.7</td>\n",
       "      <td>PB.7.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>bc1002</td>\n",
       "      <td>exon</td>\n",
       "      <td>2133698</td>\n",
       "      <td>2138792</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PB.7</td>\n",
       "      <td>PB.7.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>bc1002</td>\n",
       "      <td>exon</td>\n",
       "      <td>2139448</td>\n",
       "      <td>2139623</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PB.7</td>\n",
       "      <td>PB.7.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>172888</th>\n",
       "      <td>chrX</td>\n",
       "      <td>bc1002</td>\n",
       "      <td>exon</td>\n",
       "      <td>154364525</td>\n",
       "      <td>154364631</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PB.3112</td>\n",
       "      <td>PB.3112.38</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>172889</th>\n",
       "      <td>chrX</td>\n",
       "      <td>bc1002</td>\n",
       "      <td>transcript</td>\n",
       "      <td>154529886</td>\n",
       "      <td>154533308</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PB.3113</td>\n",
       "      <td>PB.3113.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>172890</th>\n",
       "      <td>chrX</td>\n",
       "      <td>bc1002</td>\n",
       "      <td>exon</td>\n",
       "      <td>154529886</td>\n",
       "      <td>154532280</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PB.3113</td>\n",
       "      <td>PB.3113.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>172891</th>\n",
       "      <td>chrX</td>\n",
       "      <td>bc1002</td>\n",
       "      <td>exon</td>\n",
       "      <td>154532385</td>\n",
       "      <td>154532462</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PB.3113</td>\n",
       "      <td>PB.3113.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>172892</th>\n",
       "      <td>chrX</td>\n",
       "      <td>bc1002</td>\n",
       "      <td>exon</td>\n",
       "      <td>154532566</td>\n",
       "      <td>154533308</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PB.3113</td>\n",
       "      <td>PB.3113.1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>172893 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+------------+--------------+-----------+-----------+------------+--------------+------------+------------+-----------------+\n",
       "| Chromosome   | Source     | Feature      | Start     | End       | Score      | Strand       | Frame      | gene_id    | transcript_id   |\n",
       "| (category)   | (object)   | (category)   | (int32)   | (int32)   | (object)   | (category)   | (object)   | (object)   | (object)        |\n",
       "|--------------+------------+--------------+-----------+-----------+------------+--------------+------------+------------+-----------------|\n",
       "| chr1         | bc1002     | transcript   | 854801    | 859446    | .          | +            | .          | PB.2       | PB.2.1          |\n",
       "| chr1         | bc1002     | exon         | 854801    | 859446    | .          | +            | .          | PB.2       | PB.2.1          |\n",
       "| chr1         | bc1002     | transcript   | 2133698   | 2139623   | .          | +            | .          | PB.7       | PB.7.1          |\n",
       "| chr1         | bc1002     | exon         | 2133698   | 2138792   | .          | +            | .          | PB.7       | PB.7.1          |\n",
       "| ...          | ...        | ...          | ...       | ...       | ...        | ...          | ...        | ...        | ...             |\n",
       "| chrX         | bc1002     | transcript   | 154529886 | 154533308 | .          | -            | .          | PB.3113    | PB.3113.1       |\n",
       "| chrX         | bc1002     | exon         | 154529886 | 154532280 | .          | -            | .          | PB.3113    | PB.3113.1       |\n",
       "| chrX         | bc1002     | exon         | 154532385 | 154532462 | .          | -            | .          | PB.3113    | PB.3113.1       |\n",
       "| chrX         | bc1002     | exon         | 154532566 | 154533308 | .          | -            | .          | PB.3113    | PB.3113.1       |\n",
       "+--------------+------------+--------------+-----------+-----------+------------+--------------+------------+------------+-----------------+\n",
       "Stranded PyRanges object has 172,893 rows and 10 columns from 24 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand."
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bc1002_annotation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "9df736f8-3bf8-450d-96da-1f0d4a801a0e",
   "metadata": {},
   "outputs": [],
   "source": [
    "merged_annotation = reduce(lambda i, j: merge_ranges(i, j), annotations)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "b26d4b1f-3ef1-43a6-b4b5-8fb3a102a148",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>bc1002</td>\n",
       "      <td>transcript</td>\n",
       "      <td>854801</td>\n",
       "      <td>859446</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PB.2</td>\n",
       "      <td>PB.2.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>bc1002</td>\n",
       "      <td>exon</td>\n",
       "      <td>854801</td>\n",
       "      <td>859446</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PB.2</td>\n",
       "      <td>PB.2.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>bc1002</td>\n",
       "      <td>transcript</td>\n",
       "      <td>2133698</td>\n",
       "      <td>2139623</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PB.7</td>\n",
       "      <td>PB.7.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>bc1002</td>\n",
       "      <td>exon</td>\n",
       "      <td>2133698</td>\n",
       "      <td>2138792</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PB.7</td>\n",
       "      <td>PB.7.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>bc1002</td>\n",
       "      <td>exon</td>\n",
       "      <td>2139448</td>\n",
       "      <td>2139623</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PB.7</td>\n",
       "      <td>PB.7.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2379118</th>\n",
       "      <td>chrY</td>\n",
       "      <td>bc1012</td>\n",
       "      <td>exon</td>\n",
       "      <td>19741317</td>\n",
       "      <td>19741488</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PB.3466</td>\n",
       "      <td>PB.3466.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2379119</th>\n",
       "      <td>chrY</td>\n",
       "      <td>bc1012</td>\n",
       "      <td>exon</td>\n",
       "      <td>19741734</td>\n",
       "      <td>19741857</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PB.3466</td>\n",
       "      <td>PB.3466.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2379120</th>\n",
       "      <td>chrY</td>\n",
       "      <td>bc1012</td>\n",
       "      <td>exon</td>\n",
       "      <td>19743161</td>\n",
       "      <td>19743239</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PB.3466</td>\n",
       "      <td>PB.3466.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2379121</th>\n",
       "      <td>chrY</td>\n",
       "      <td>bc1012</td>\n",
       "      <td>exon</td>\n",
       "      <td>19744384</td>\n",
       "      <td>19744553</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PB.3466</td>\n",
       "      <td>PB.3466.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2379122</th>\n",
       "      <td>chrY</td>\n",
       "      <td>bc1012</td>\n",
       "      <td>exon</td>\n",
       "      <td>19744670</td>\n",
       "      <td>19744714</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PB.3466</td>\n",
       "      <td>PB.3466.5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2379123 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+------------+--------------+-----------+-----------+------------+--------------+------------+------------+-----------------+\n",
       "| Chromosome   | Source     | Feature      | Start     | End       | Score      | Strand       | Frame      | gene_id    | transcript_id   |\n",
       "| (category)   | (object)   | (category)   | (int32)   | (int32)   | (object)   | (category)   | (object)   | (object)   | (object)        |\n",
       "|--------------+------------+--------------+-----------+-----------+------------+--------------+------------+------------+-----------------|\n",
       "| chr1         | bc1002     | transcript   | 854801    | 859446    | .          | +            | .          | PB.2       | PB.2.1          |\n",
       "| chr1         | bc1002     | exon         | 854801    | 859446    | .          | +            | .          | PB.2       | PB.2.1          |\n",
       "| chr1         | bc1002     | transcript   | 2133698   | 2139623   | .          | +            | .          | PB.7       | PB.7.1          |\n",
       "| chr1         | bc1002     | exon         | 2133698   | 2138792   | .          | +            | .          | PB.7       | PB.7.1          |\n",
       "| ...          | ...        | ...          | ...       | ...       | ...        | ...          | ...        | ...        | ...             |\n",
       "| chrY         | bc1012     | exon         | 19741734  | 19741857  | .          | -            | .          | PB.3466    | PB.3466.5       |\n",
       "| chrY         | bc1012     | exon         | 19743161  | 19743239  | .          | -            | .          | PB.3466    | PB.3466.5       |\n",
       "| chrY         | bc1012     | exon         | 19744384  | 19744553  | .          | -            | .          | PB.3466    | PB.3466.5       |\n",
       "| chrY         | bc1012     | exon         | 19744670  | 19744714  | .          | -            | .          | PB.3466    | PB.3466.5       |\n",
       "+--------------+------------+--------------+-----------+-----------+------------+--------------+------------+------------+-----------------+\n",
       "Stranded PyRanges object has 2,379,123 rows and 10 columns from 35 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand."
      ]
     },
     "execution_count": 61,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged_annotation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "ca66cb85-d930-4234-a5ab-a0b32a9e4579",
   "metadata": {},
   "outputs": [],
   "source": [
    "merged_annotation = merged_annotation.cluster(nb_cpu=6)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "f9b87bca-6878-431c-85d7-1da0e7097afb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>Cluster</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>bc1003</td>\n",
       "      <td>transcript</td>\n",
       "      <td>827658</td>\n",
       "      <td>857173</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PB.2</td>\n",
       "      <td>PB.2.1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>bc1003</td>\n",
       "      <td>exon</td>\n",
       "      <td>827658</td>\n",
       "      <td>827775</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PB.2</td>\n",
       "      <td>PB.2.1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>bc1003</td>\n",
       "      <td>exon</td>\n",
       "      <td>829002</td>\n",
       "      <td>829104</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PB.2</td>\n",
       "      <td>PB.2.1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>bc1004</td>\n",
       "      <td>transcript</td>\n",
       "      <td>833846</td>\n",
       "      <td>838936</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PB.3</td>\n",
       "      <td>PB.3.1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>bc1004</td>\n",
       "      <td>exon</td>\n",
       "      <td>833846</td>\n",
       "      <td>838936</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PB.3</td>\n",
       "      <td>PB.3.1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2379118</th>\n",
       "      <td>chrY</td>\n",
       "      <td>bc1012</td>\n",
       "      <td>exon</td>\n",
       "      <td>19744670</td>\n",
       "      <td>19744723</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PB.3466</td>\n",
       "      <td>PB.3466.4</td>\n",
       "      <td>3836</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2379119</th>\n",
       "      <td>chrY</td>\n",
       "      <td>bc1012</td>\n",
       "      <td>exon</td>\n",
       "      <td>19744670</td>\n",
       "      <td>19744723</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PB.3466</td>\n",
       "      <td>PB.3466.2</td>\n",
       "      <td>3836</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2379120</th>\n",
       "      <td>chrY</td>\n",
       "      <td>bc1012</td>\n",
       "      <td>exon</td>\n",
       "      <td>19744670</td>\n",
       "      <td>19744726</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PB.3466</td>\n",
       "      <td>PB.3466.3</td>\n",
       "      <td>3836</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2379121</th>\n",
       "      <td>chrY</td>\n",
       "      <td>bc1012</td>\n",
       "      <td>exon</td>\n",
       "      <td>19744670</td>\n",
       "      <td>19744726</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PB.3466</td>\n",
       "      <td>PB.3466.1</td>\n",
       "      <td>3836</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2379122</th>\n",
       "      <td>chrY</td>\n",
       "      <td>bc1012</td>\n",
       "      <td>exon</td>\n",
       "      <td>19744670</td>\n",
       "      <td>19744714</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PB.3466</td>\n",
       "      <td>PB.3466.5</td>\n",
       "      <td>3836</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2379123 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+------------+--------------+-----------+-----------+------------+--------------+------------+------------+-----------------+-----------+\n",
       "| Chromosome   | Source     | Feature      | Start     | End       | Score      | Strand       | Frame      | gene_id    | transcript_id   | Cluster   |\n",
       "| (category)   | (object)   | (category)   | (int32)   | (int32)   | (object)   | (category)   | (object)   | (object)   | (object)        | (int32)   |\n",
       "|--------------+------------+--------------+-----------+-----------+------------+--------------+------------+------------+-----------------+-----------|\n",
       "| chr1         | bc1003     | transcript   | 827658    | 857173    | .          | +            | .          | PB.2       | PB.2.1          | 1         |\n",
       "| chr1         | bc1003     | exon         | 827658    | 827775    | .          | +            | .          | PB.2       | PB.2.1          | 1         |\n",
       "| chr1         | bc1003     | exon         | 829002    | 829104    | .          | +            | .          | PB.2       | PB.2.1          | 1         |\n",
       "| chr1         | bc1004     | transcript   | 833846    | 838936    | .          | +            | .          | PB.3       | PB.3.1          | 1         |\n",
       "| ...          | ...        | ...          | ...       | ...       | ...        | ...          | ...        | ...        | ...             | ...       |\n",
       "| chrY         | bc1012     | exon         | 19744670  | 19744723  | .          | -            | .          | PB.3466    | PB.3466.2       | 3836      |\n",
       "| chrY         | bc1012     | exon         | 19744670  | 19744726  | .          | -            | .          | PB.3466    | PB.3466.3       | 3836      |\n",
       "| chrY         | bc1012     | exon         | 19744670  | 19744726  | .          | -            | .          | PB.3466    | PB.3466.1       | 3836      |\n",
       "| chrY         | bc1012     | exon         | 19744670  | 19744714  | .          | -            | .          | PB.3466    | PB.3466.5       | 3836      |\n",
       "+--------------+------------+--------------+-----------+-----------+------------+--------------+------------+------------+-----------------+-----------+\n",
       "Stranded PyRanges object has 2,379,123 rows and 11 columns from 35 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand."
      ]
     },
     "execution_count": 63,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged_annotation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "b751c871-3bc9-4c85-ac07-7514de5b4441",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>Cluster</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>bc1003</td>\n",
       "      <td>transcript</td>\n",
       "      <td>827658</td>\n",
       "      <td>857173</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PB.2</td>\n",
       "      <td>PB.2.1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>bc1003</td>\n",
       "      <td>exon</td>\n",
       "      <td>827658</td>\n",
       "      <td>827775</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PB.2</td>\n",
       "      <td>PB.2.1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>bc1003</td>\n",
       "      <td>exon</td>\n",
       "      <td>829002</td>\n",
       "      <td>829104</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PB.2</td>\n",
       "      <td>PB.2.1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>bc1004</td>\n",
       "      <td>transcript</td>\n",
       "      <td>833846</td>\n",
       "      <td>838936</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PB.3</td>\n",
       "      <td>PB.3.1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>bc1004</td>\n",
       "      <td>exon</td>\n",
       "      <td>833846</td>\n",
       "      <td>838936</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PB.3</td>\n",
       "      <td>PB.3.1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr1</td>\n",
       "      <td>bc1005</td>\n",
       "      <td>transcript</td>\n",
       "      <td>834239</td>\n",
       "      <td>838441</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PB.1</td>\n",
       "      <td>PB.1.1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr1</td>\n",
       "      <td>bc1005</td>\n",
       "      <td>exon</td>\n",
       "      <td>834239</td>\n",
       "      <td>838441</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PB.1</td>\n",
       "      <td>PB.1.1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chr1</td>\n",
       "      <td>bc1003</td>\n",
       "      <td>exon</td>\n",
       "      <td>851926</td>\n",
       "      <td>852110</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PB.2</td>\n",
       "      <td>PB.2.1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>chr1</td>\n",
       "      <td>bc1003</td>\n",
       "      <td>exon</td>\n",
       "      <td>853390</td>\n",
       "      <td>857173</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PB.2</td>\n",
       "      <td>PB.2.1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>chr1</td>\n",
       "      <td>bc1002</td>\n",
       "      <td>transcript</td>\n",
       "      <td>854801</td>\n",
       "      <td>859446</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PB.2</td>\n",
       "      <td>PB.2.1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>chr1</td>\n",
       "      <td>bc1002</td>\n",
       "      <td>exon</td>\n",
       "      <td>854801</td>\n",
       "      <td>859446</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PB.2</td>\n",
       "      <td>PB.2.1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+------------+--------------+-----------+-----------+------------+--------------+------------+------------+-----------------+-----------+\n",
       "| Chromosome   | Source     | Feature      | Start     | End       | Score      | Strand       | Frame      | gene_id    | transcript_id   | Cluster   |\n",
       "| (category)   | (object)   | (category)   | (int32)   | (int32)   | (object)   | (category)   | (object)   | (object)   | (object)        | (int32)   |\n",
       "|--------------+------------+--------------+-----------+-----------+------------+--------------+------------+------------+-----------------+-----------|\n",
       "| chr1         | bc1003     | transcript   | 827658    | 857173    | .          | +            | .          | PB.2       | PB.2.1          | 1         |\n",
       "| chr1         | bc1003     | exon         | 827658    | 827775    | .          | +            | .          | PB.2       | PB.2.1          | 1         |\n",
       "| chr1         | bc1003     | exon         | 829002    | 829104    | .          | +            | .          | PB.2       | PB.2.1          | 1         |\n",
       "| chr1         | bc1004     | transcript   | 833846    | 838936    | .          | +            | .          | PB.3       | PB.3.1          | 1         |\n",
       "| ...          | ...        | ...          | ...       | ...       | ...        | ...          | ...        | ...        | ...             | ...       |\n",
       "| chr1         | bc1003     | exon         | 851926    | 852110    | .          | +            | .          | PB.2       | PB.2.1          | 1         |\n",
       "| chr1         | bc1003     | exon         | 853390    | 857173    | .          | +            | .          | PB.2       | PB.2.1          | 1         |\n",
       "| chr1         | bc1002     | transcript   | 854801    | 859446    | .          | +            | .          | PB.2       | PB.2.1          | 1         |\n",
       "| chr1         | bc1002     | exon         | 854801    | 859446    | .          | +            | .          | PB.2       | PB.2.1          | 1         |\n",
       "+--------------+------------+--------------+-----------+-----------+------------+--------------+------------+------------+-----------------+-----------+\n",
       "Stranded PyRanges object has 11 rows and 11 columns from 1 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand."
      ]
     },
     "execution_count": 64,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged_annotation.subset(lambda x: x.Cluster == 1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "932472bf-4d38-484a-9d75-5787f42ff3e2",
   "metadata": {},
   "outputs": [],
   "source": [
    "from multiprocessing import cpu_count"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "fb3d1c10-53ea-4908-8ddf-96a987361129",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "12"
      ]
     },
     "execution_count": 66,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cpu_count()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
