{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "fdd53d79-b012-43be-80d9-5e646aaf6ccb",
   "metadata": {},
   "source": [
    "`chain_samples()` function in [cDNA_Cupcake](https://github.com/Magdoll/cDNA_Cupcake) crashes with an out-of-memory error on some of my samples.  Try as I might, I've not been able to decode exactly what and why it does what it does (why does it write *everything* to a file and read it back in?).  However, I found that the [pyranges](https://github.com/biocore-ntnu/pyranges) module has a `cluster` function that will give a common id to overlapping sequences.  Need to see if that will work."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "ae86fc95-baa2-4ff7-83b7-5a4382d72c13",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pyranges as pr\n",
    "import pandas as pd\n",
    "from pathlib import Path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "e1adba65-c117-4f09-97fd-2521f9d7c395",
   "metadata": {},
   "outputs": [],
   "source": [
    "bc1001_gff = Path(\"test_data/bc1001/repaired.sample.gff\")\n",
    "bc1002_gff = Path(\"test_data/bc1002/repaired.sample.gff\")\n",
    "gencode_v37_gff = Path(\"test_data/gencode.v37.annotation.gtf.gz\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "bd938b6e-4886-4e36-9687-243530b4da49",
   "metadata": {},
   "outputs": [],
   "source": [
    "bc1001_annot = pr.read_gff(bc1001_gff)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "4f8f2eb7-fa41-4559-8d5e-b9a7f3a41892",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>PacBio</td>\n",
       "      <td>transcript</td>\n",
       "      <td>1389718</td>\n",
       "      <td>1394777</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PB.3</td>\n",
       "      <td>PB.3.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>PacBio</td>\n",
       "      <td>exon</td>\n",
       "      <td>1389718</td>\n",
       "      <td>1390368</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PB.3</td>\n",
       "      <td>PB.3.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>PacBio</td>\n",
       "      <td>exon</td>\n",
       "      <td>1390455</td>\n",
       "      <td>1390565</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PB.3</td>\n",
       "      <td>PB.3.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>PacBio</td>\n",
       "      <td>exon</td>\n",
       "      <td>1390767</td>\n",
       "      <td>1394777</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PB.3</td>\n",
       "      <td>PB.3.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>PacBio</td>\n",
       "      <td>transcript</td>\n",
       "      <td>2133700</td>\n",
       "      <td>2139625</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PB.7</td>\n",
       "      <td>PB.7.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>208497</th>\n",
       "      <td>chrX</td>\n",
       "      <td>PacBio</td>\n",
       "      <td>exon</td>\n",
       "      <td>154789953</td>\n",
       "      <td>154790022</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PB.3323</td>\n",
       "      <td>PB.3323.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>208498</th>\n",
       "      <td>chrX</td>\n",
       "      <td>PacBio</td>\n",
       "      <td>exon</td>\n",
       "      <td>154790982</td>\n",
       "      <td>154791068</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PB.3323</td>\n",
       "      <td>PB.3323.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>208499</th>\n",
       "      <td>chrX</td>\n",
       "      <td>PacBio</td>\n",
       "      <td>exon</td>\n",
       "      <td>154791768</td>\n",
       "      <td>154791847</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PB.3323</td>\n",
       "      <td>PB.3323.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>208500</th>\n",
       "      <td>chrX</td>\n",
       "      <td>PacBio</td>\n",
       "      <td>exon</td>\n",
       "      <td>154792141</td>\n",
       "      <td>154792285</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PB.3323</td>\n",
       "      <td>PB.3323.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>208501</th>\n",
       "      <td>chrX</td>\n",
       "      <td>PacBio</td>\n",
       "      <td>exon</td>\n",
       "      <td>154805271</td>\n",
       "      <td>154805472</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PB.3323</td>\n",
       "      <td>PB.3323.1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>208502 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+------------+--------------+-----------+-----------+------------+--------------+------------+------------+-----------------+\n",
       "| Chromosome   | Source     | Feature      | Start     | End       | Score      | Strand       | Frame      | gene_id    | transcript_id   |\n",
       "| (category)   | (object)   | (category)   | (int32)   | (int32)   | (object)   | (category)   | (object)   | (object)   | (object)        |\n",
       "|--------------+------------+--------------+-----------+-----------+------------+--------------+------------+------------+-----------------|\n",
       "| chr1         | PacBio     | transcript   | 1389718   | 1394777   | .          | +            | .          | PB.3       | PB.3.1          |\n",
       "| chr1         | PacBio     | exon         | 1389718   | 1390368   | .          | +            | .          | PB.3       | PB.3.1          |\n",
       "| chr1         | PacBio     | exon         | 1390455   | 1390565   | .          | +            | .          | PB.3       | PB.3.1          |\n",
       "| chr1         | PacBio     | exon         | 1390767   | 1394777   | .          | +            | .          | PB.3       | PB.3.1          |\n",
       "| ...          | ...        | ...          | ...       | ...       | ...        | ...          | ...        | ...        | ...             |\n",
       "| chrX         | PacBio     | exon         | 154790982 | 154791068 | .          | -            | .          | PB.3323    | PB.3323.1       |\n",
       "| chrX         | PacBio     | exon         | 154791768 | 154791847 | .          | -            | .          | PB.3323    | PB.3323.1       |\n",
       "| chrX         | PacBio     | exon         | 154792141 | 154792285 | .          | -            | .          | PB.3323    | PB.3323.1       |\n",
       "| chrX         | PacBio     | exon         | 154805271 | 154805472 | .          | -            | .          | PB.3323    | PB.3323.1       |\n",
       "+--------------+------------+--------------+-----------+-----------+------------+--------------+------------+------------+-----------------+\n",
       "Stranded PyRanges object has 208,502 rows and 10 columns from 23 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand."
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bc1001_annot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "cf96b841-e85e-42ed-ae5f-2a416c9bdeba",
   "metadata": {},
   "outputs": [],
   "source": [
    "bc1002_annot = pr.read_gff(bc1002_gff)\n",
    "# gencode = pr.read_gff(gencode_v37_gff)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "fcc22590-cd15-4963-8cd6-ed68772a5430",
   "metadata": {},
   "outputs": [],
   "source": [
    "# gencode"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "74b1f456-335d-4187-8c1d-896003d93917",
   "metadata": {},
   "outputs": [],
   "source": [
    "# gencode.subset(lambda x: x[\"Feature\"].isin((\"transcript\", \"exon\")))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "ee3c3ac7-242e-425b-a13e-dd3cd81b25e3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# gencode_expressed = gencode.subset(lambda x: x[\"Feature\"].isin((\"transcript\", \"exon\")))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "13a112b1-5118-42cd-9fb6-c6a8efea0a15",
   "metadata": {},
   "outputs": [],
   "source": [
    "bc1001_sample_name = pd.Series([\"bc1001\" for _ in range(len(bc1001_annot))], name=\"sample_source\")\n",
    "bc1001_annot = bc1001_annot.insert(bc1001_sample_name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "fcaef734-c2b7-4f3e-98fe-89877ab4a5da",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>sample_source</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>PacBio</td>\n",
       "      <td>transcript</td>\n",
       "      <td>1389718</td>\n",
       "      <td>1394777</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PB.3</td>\n",
       "      <td>PB.3.1</td>\n",
       "      <td>bc1001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>PacBio</td>\n",
       "      <td>exon</td>\n",
       "      <td>1389718</td>\n",
       "      <td>1390368</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PB.3</td>\n",
       "      <td>PB.3.1</td>\n",
       "      <td>bc1001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>PacBio</td>\n",
       "      <td>exon</td>\n",
       "      <td>1390455</td>\n",
       "      <td>1390565</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PB.3</td>\n",
       "      <td>PB.3.1</td>\n",
       "      <td>bc1001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>PacBio</td>\n",
       "      <td>exon</td>\n",
       "      <td>1390767</td>\n",
       "      <td>1394777</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PB.3</td>\n",
       "      <td>PB.3.1</td>\n",
       "      <td>bc1001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>PacBio</td>\n",
       "      <td>transcript</td>\n",
       "      <td>2133700</td>\n",
       "      <td>2139625</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PB.7</td>\n",
       "      <td>PB.7.1</td>\n",
       "      <td>bc1001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr1</td>\n",
       "      <td>PacBio</td>\n",
       "      <td>exon</td>\n",
       "      <td>2133700</td>\n",
       "      <td>2139625</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PB.7</td>\n",
       "      <td>PB.7.1</td>\n",
       "      <td>bc1001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr1</td>\n",
       "      <td>PacBio</td>\n",
       "      <td>transcript</td>\n",
       "      <td>2245074</td>\n",
       "      <td>2280214</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PB.8</td>\n",
       "      <td>PB.8.1</td>\n",
       "      <td>bc1001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chr1</td>\n",
       "      <td>PacBio</td>\n",
       "      <td>exon</td>\n",
       "      <td>2245074</td>\n",
       "      <td>2245959</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PB.8</td>\n",
       "      <td>PB.8.1</td>\n",
       "      <td>bc1001</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+------------+--------------+-----------+-----------+------------+--------------+------------+------------+-----------------+-----------------+\n",
       "| Chromosome   | Source     | Feature      |     Start |       End | Score      | Strand       | Frame      | gene_id    | transcript_id   | sample_source   |\n",
       "| (category)   | (object)   | (category)   |   (int32) |   (int32) | (object)   | (category)   | (object)   | (object)   | (object)        | (object)        |\n",
       "|--------------+------------+--------------+-----------+-----------+------------+--------------+------------+------------+-----------------+-----------------|\n",
       "| chr1         | PacBio     | transcript   |   1389718 |   1394777 | .          | +            | .          | PB.3       | PB.3.1          | bc1001          |\n",
       "| chr1         | PacBio     | exon         |   1389718 |   1390368 | .          | +            | .          | PB.3       | PB.3.1          | bc1001          |\n",
       "| chr1         | PacBio     | exon         |   1390455 |   1390565 | .          | +            | .          | PB.3       | PB.3.1          | bc1001          |\n",
       "| chr1         | PacBio     | exon         |   1390767 |   1394777 | .          | +            | .          | PB.3       | PB.3.1          | bc1001          |\n",
       "| chr1         | PacBio     | transcript   |   2133700 |   2139625 | .          | +            | .          | PB.7       | PB.7.1          | bc1001          |\n",
       "| chr1         | PacBio     | exon         |   2133700 |   2139625 | .          | +            | .          | PB.7       | PB.7.1          | bc1001          |\n",
       "| chr1         | PacBio     | transcript   |   2245074 |   2280214 | .          | +            | .          | PB.8       | PB.8.1          | bc1001          |\n",
       "| chr1         | PacBio     | exon         |   2245074 |   2245959 | .          | +            | .          | PB.8       | PB.8.1          | bc1001          |\n",
       "+--------------+------------+--------------+-----------+-----------+------------+--------------+------------+------------+-----------------+-----------------+\n",
       "Stranded PyRanges object has 8 rows and 11 columns from 1 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand."
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bc1001_annot.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "8c52d7e7-bc0e-400e-9a70-3ec077fe15ef",
   "metadata": {},
   "outputs": [],
   "source": [
    "bc1002_sample_name = pd.Series([\"bc1002\" for _ in range(len(bc1002_annot))], name=\"sample_source\")\n",
    "bc1002_annot = bc1002_annot.insert(bc1002_sample_name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "f1abc51b-4227-469c-82e0-c51902e421c5",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2021-04-30 15:37:30,790\tINFO services.py:1171 -- View the Ray dashboard at \u001b[1m\u001b[32mhttp://127.0.0.1:8265\u001b[39m\u001b[22m\n"
     ]
    }
   ],
   "source": [
    "bc1001_1002 = bc1001_annot.join(bc1002_annot, how=\"left\", nb_cpu=4)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1020ea85-3ed9-4dfb-827d-8437295526e5",
   "metadata": {},
   "source": [
    "Note: one issue here is that `join()` has to perform some sort of table join, it won't simply bind rows.  This necessarily means we are going to lose samples that are not present in both PyRanges objects. \n",
    "\n",
    "We could probably deal with this by performing multiple joins or, if nothing else, one can probably get around this by taking all the underlying dataframes, merging those together, converting to a `dict`, and then converting that back into a `PyRanges` object"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "bdbef5dc-67ff-4f8a-b6cb-9ce9ddfa97b7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>...</th>\n",
       "      <th>Source_b</th>\n",
       "      <th>Feature_b</th>\n",
       "      <th>Start_b</th>\n",
       "      <th>End_b</th>\n",
       "      <th>Score_b</th>\n",
       "      <th>Strand_b</th>\n",
       "      <th>Frame_b</th>\n",
       "      <th>gene_id_b</th>\n",
       "      <th>transcript_id_b</th>\n",
       "      <th>sample_source_b</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>PacBio</td>\n",
       "      <td>transcript</td>\n",
       "      <td>1389718</td>\n",
       "      <td>1394777</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PB.3</td>\n",
       "      <td>PB.3.1</td>\n",
       "      <td>...</td>\n",
       "      <td>PacBio</td>\n",
       "      <td>transcript</td>\n",
       "      <td>1385709</td>\n",
       "      <td>1399339</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PB.4</td>\n",
       "      <td>PB.4.1</td>\n",
       "      <td>bc1002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>PacBio</td>\n",
       "      <td>transcript</td>\n",
       "      <td>1389718</td>\n",
       "      <td>1394777</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PB.3</td>\n",
       "      <td>PB.3.1</td>\n",
       "      <td>...</td>\n",
       "      <td>PacBio</td>\n",
       "      <td>transcript</td>\n",
       "      <td>1386500</td>\n",
       "      <td>1399317</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PB.4</td>\n",
       "      <td>PB.4.2</td>\n",
       "      <td>bc1002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>PacBio</td>\n",
       "      <td>transcript</td>\n",
       "      <td>1389718</td>\n",
       "      <td>1394777</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PB.3</td>\n",
       "      <td>PB.3.1</td>\n",
       "      <td>...</td>\n",
       "      <td>PacBio</td>\n",
       "      <td>transcript</td>\n",
       "      <td>1386504</td>\n",
       "      <td>1399338</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PB.4</td>\n",
       "      <td>PB.4.3</td>\n",
       "      <td>bc1002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>PacBio</td>\n",
       "      <td>transcript</td>\n",
       "      <td>1389718</td>\n",
       "      <td>1394777</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PB.3</td>\n",
       "      <td>PB.3.1</td>\n",
       "      <td>...</td>\n",
       "      <td>PacBio</td>\n",
       "      <td>transcript</td>\n",
       "      <td>1386705</td>\n",
       "      <td>1399338</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PB.4</td>\n",
       "      <td>PB.4.4</td>\n",
       "      <td>bc1002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>PacBio</td>\n",
       "      <td>transcript</td>\n",
       "      <td>1389718</td>\n",
       "      <td>1394777</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PB.3</td>\n",
       "      <td>PB.3.1</td>\n",
       "      <td>...</td>\n",
       "      <td>PacBio</td>\n",
       "      <td>transcript</td>\n",
       "      <td>1386709</td>\n",
       "      <td>1399328</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PB.4</td>\n",
       "      <td>PB.4.5</td>\n",
       "      <td>bc1002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr1</td>\n",
       "      <td>PacBio</td>\n",
       "      <td>transcript</td>\n",
       "      <td>1389718</td>\n",
       "      <td>1394777</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PB.3</td>\n",
       "      <td>PB.3.1</td>\n",
       "      <td>...</td>\n",
       "      <td>PacBio</td>\n",
       "      <td>transcript</td>\n",
       "      <td>1386710</td>\n",
       "      <td>1399335</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PB.4</td>\n",
       "      <td>PB.4.7</td>\n",
       "      <td>bc1002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr1</td>\n",
       "      <td>PacBio</td>\n",
       "      <td>transcript</td>\n",
       "      <td>1389718</td>\n",
       "      <td>1394777</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PB.3</td>\n",
       "      <td>PB.3.1</td>\n",
       "      <td>...</td>\n",
       "      <td>PacBio</td>\n",
       "      <td>transcript</td>\n",
       "      <td>1386710</td>\n",
       "      <td>1399317</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PB.4</td>\n",
       "      <td>PB.4.6</td>\n",
       "      <td>bc1002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chr1</td>\n",
       "      <td>PacBio</td>\n",
       "      <td>transcript</td>\n",
       "      <td>1389718</td>\n",
       "      <td>1394777</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PB.3</td>\n",
       "      <td>PB.3.1</td>\n",
       "      <td>...</td>\n",
       "      <td>PacBio</td>\n",
       "      <td>transcript</td>\n",
       "      <td>1387343</td>\n",
       "      <td>1399317</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PB.4</td>\n",
       "      <td>PB.4.8</td>\n",
       "      <td>bc1002</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>8 rows × 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+------------+------------+-----------+-----------+------------+--------------+------------+------------+-----------------+-----------------+------------+-------------+-----------+-----------+------------+--------------+------------+-------+\n",
       "| Chromosome   | Source     | Feature    |     Start |       End | Score      | Strand       | Frame      | gene_id    | transcript_id   | sample_source   | Source_b   | Feature_b   |   Start_b |     End_b | Score_b    | Strand_b     | Frame_b    | +3    |\n",
       "| (category)   | (object)   | (object)   |   (int64) |   (int64) | (object)   | (category)   | (object)   | (object)   | (object)        | (object)        | (object)   | (object)    |   (int64) |   (int64) | (object)   | (category)   | (object)   | ...   |\n",
       "|--------------+------------+------------+-----------+-----------+------------+--------------+------------+------------+-----------------+-----------------+------------+-------------+-----------+-----------+------------+--------------+------------+-------|\n",
       "| chr1         | PacBio     | transcript |   1389718 |   1394777 | .          | +            | .          | PB.3       | PB.3.1          | bc1001          | PacBio     | transcript  |   1385709 |   1399339 | .          | -            | .          | ...   |\n",
       "| chr1         | PacBio     | transcript |   1389718 |   1394777 | .          | +            | .          | PB.3       | PB.3.1          | bc1001          | PacBio     | transcript  |   1386500 |   1399317 | .          | -            | .          | ...   |\n",
       "| chr1         | PacBio     | transcript |   1389718 |   1394777 | .          | +            | .          | PB.3       | PB.3.1          | bc1001          | PacBio     | transcript  |   1386504 |   1399338 | .          | -            | .          | ...   |\n",
       "| chr1         | PacBio     | transcript |   1389718 |   1394777 | .          | +            | .          | PB.3       | PB.3.1          | bc1001          | PacBio     | transcript  |   1386705 |   1399338 | .          | -            | .          | ...   |\n",
       "| chr1         | PacBio     | transcript |   1389718 |   1394777 | .          | +            | .          | PB.3       | PB.3.1          | bc1001          | PacBio     | transcript  |   1386709 |   1399328 | .          | -            | .          | ...   |\n",
       "| chr1         | PacBio     | transcript |   1389718 |   1394777 | .          | +            | .          | PB.3       | PB.3.1          | bc1001          | PacBio     | transcript  |   1386710 |   1399335 | .          | -            | .          | ...   |\n",
       "| chr1         | PacBio     | transcript |   1389718 |   1394777 | .          | +            | .          | PB.3       | PB.3.1          | bc1001          | PacBio     | transcript  |   1386710 |   1399317 | .          | -            | .          | ...   |\n",
       "| chr1         | PacBio     | transcript |   1389718 |   1394777 | .          | +            | .          | PB.3       | PB.3.1          | bc1001          | PacBio     | transcript  |   1387343 |   1399317 | .          | -            | .          | ...   |\n",
       "+--------------+------------+------------+-----------+-----------+------------+--------------+------------+------------+-----------------+-----------------+------------+-------------+-----------+-----------+------------+--------------+------------+-------+\n",
       "Stranded PyRanges object has 8 rows and 21 columns from 1 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "3 hidden columns: gene_id_b, transcript_id_b, sample_source_b"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bc1001_1002.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "d1cba64d-b5e0-46f5-bedd-7d18783a207f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>...</th>\n",
       "      <th>Feature_b</th>\n",
       "      <th>Start_b</th>\n",
       "      <th>End_b</th>\n",
       "      <th>Score_b</th>\n",
       "      <th>Strand_b</th>\n",
       "      <th>Frame_b</th>\n",
       "      <th>gene_id_b</th>\n",
       "      <th>transcript_id_b</th>\n",
       "      <th>sample_source_b</th>\n",
       "      <th>Cluster</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>PacBio</td>\n",
       "      <td>transcript</td>\n",
       "      <td>1389718</td>\n",
       "      <td>1394777</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PB.3</td>\n",
       "      <td>PB.3.1</td>\n",
       "      <td>...</td>\n",
       "      <td>transcript</td>\n",
       "      <td>1385709</td>\n",
       "      <td>1399339</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PB.4</td>\n",
       "      <td>PB.4.1</td>\n",
       "      <td>bc1002</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>PacBio</td>\n",
       "      <td>transcript</td>\n",
       "      <td>1389718</td>\n",
       "      <td>1394777</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PB.3</td>\n",
       "      <td>PB.3.1</td>\n",
       "      <td>...</td>\n",
       "      <td>exon</td>\n",
       "      <td>1390765</td>\n",
       "      <td>1392803</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PB.4</td>\n",
       "      <td>PB.4.8</td>\n",
       "      <td>bc1002</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>PacBio</td>\n",
       "      <td>transcript</td>\n",
       "      <td>1389718</td>\n",
       "      <td>1394777</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PB.3</td>\n",
       "      <td>PB.3.1</td>\n",
       "      <td>...</td>\n",
       "      <td>exon</td>\n",
       "      <td>1390765</td>\n",
       "      <td>1392803</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PB.4</td>\n",
       "      <td>PB.4.9</td>\n",
       "      <td>bc1002</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>PacBio</td>\n",
       "      <td>transcript</td>\n",
       "      <td>1389718</td>\n",
       "      <td>1394777</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PB.3</td>\n",
       "      <td>PB.3.1</td>\n",
       "      <td>...</td>\n",
       "      <td>exon</td>\n",
       "      <td>1393395</td>\n",
       "      <td>1393460</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PB.4</td>\n",
       "      <td>PB.4.1</td>\n",
       "      <td>bc1002</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>PacBio</td>\n",
       "      <td>transcript</td>\n",
       "      <td>1389718</td>\n",
       "      <td>1394777</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PB.3</td>\n",
       "      <td>PB.3.1</td>\n",
       "      <td>...</td>\n",
       "      <td>exon</td>\n",
       "      <td>1393395</td>\n",
       "      <td>1393460</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PB.4</td>\n",
       "      <td>PB.4.2</td>\n",
       "      <td>bc1002</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>103</th>\n",
       "      <td>chr1</td>\n",
       "      <td>PacBio</td>\n",
       "      <td>exon</td>\n",
       "      <td>1390767</td>\n",
       "      <td>1394777</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PB.3</td>\n",
       "      <td>PB.3.1</td>\n",
       "      <td>...</td>\n",
       "      <td>transcript</td>\n",
       "      <td>1387343</td>\n",
       "      <td>1399317</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PB.4</td>\n",
       "      <td>PB.4.8</td>\n",
       "      <td>bc1002</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>104</th>\n",
       "      <td>chr1</td>\n",
       "      <td>PacBio</td>\n",
       "      <td>exon</td>\n",
       "      <td>1390767</td>\n",
       "      <td>1394777</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PB.3</td>\n",
       "      <td>PB.3.1</td>\n",
       "      <td>...</td>\n",
       "      <td>transcript</td>\n",
       "      <td>1390334</td>\n",
       "      <td>1399339</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PB.4</td>\n",
       "      <td>PB.4.9</td>\n",
       "      <td>bc1002</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>105</th>\n",
       "      <td>chr1</td>\n",
       "      <td>PacBio</td>\n",
       "      <td>exon</td>\n",
       "      <td>1390767</td>\n",
       "      <td>1394777</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PB.3</td>\n",
       "      <td>PB.3.1</td>\n",
       "      <td>...</td>\n",
       "      <td>exon</td>\n",
       "      <td>1390765</td>\n",
       "      <td>1392803</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PB.4</td>\n",
       "      <td>PB.4.1</td>\n",
       "      <td>bc1002</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>106</th>\n",
       "      <td>chr1</td>\n",
       "      <td>PacBio</td>\n",
       "      <td>exon</td>\n",
       "      <td>1390767</td>\n",
       "      <td>1394777</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PB.3</td>\n",
       "      <td>PB.3.1</td>\n",
       "      <td>...</td>\n",
       "      <td>exon</td>\n",
       "      <td>1390765</td>\n",
       "      <td>1392803</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PB.4</td>\n",
       "      <td>PB.4.2</td>\n",
       "      <td>bc1002</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>107</th>\n",
       "      <td>chr1</td>\n",
       "      <td>PacBio</td>\n",
       "      <td>exon</td>\n",
       "      <td>1390767</td>\n",
       "      <td>1394777</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PB.3</td>\n",
       "      <td>PB.3.1</td>\n",
       "      <td>...</td>\n",
       "      <td>transcript</td>\n",
       "      <td>1386710</td>\n",
       "      <td>1399335</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PB.4</td>\n",
       "      <td>PB.4.7</td>\n",
       "      <td>bc1002</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>108 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+------------+------------+-----------+-----------+------------+--------------+------------+------------+-----------------+-----------------+------------+-------------+-----------+-----------+------------+--------------+------------+-------+\n",
       "| Chromosome   | Source     | Feature    | Start     | End       | Score      | Strand       | Frame      | gene_id    | transcript_id   | sample_source   | Source_b   | Feature_b   | Start_b   | End_b     | Score_b    | Strand_b     | Frame_b    | +4    |\n",
       "| (category)   | (object)   | (object)   | (int64)   | (int64)   | (object)   | (category)   | (object)   | (object)   | (object)        | (object)        | (object)   | (object)    | (int64)   | (int64)   | (object)   | (category)   | (object)   | ...   |\n",
       "|--------------+------------+------------+-----------+-----------+------------+--------------+------------+------------+-----------------+-----------------+------------+-------------+-----------+-----------+------------+--------------+------------+-------|\n",
       "| chr1         | PacBio     | transcript | 1389718   | 1394777   | .          | +            | .          | PB.3       | PB.3.1          | bc1001          | PacBio     | transcript  | 1385709   | 1399339   | .          | -            | .          | ...   |\n",
       "| chr1         | PacBio     | transcript | 1389718   | 1394777   | .          | +            | .          | PB.3       | PB.3.1          | bc1001          | PacBio     | exon        | 1390765   | 1392803   | .          | -            | .          | ...   |\n",
       "| chr1         | PacBio     | transcript | 1389718   | 1394777   | .          | +            | .          | PB.3       | PB.3.1          | bc1001          | PacBio     | exon        | 1390765   | 1392803   | .          | -            | .          | ...   |\n",
       "| chr1         | PacBio     | transcript | 1389718   | 1394777   | .          | +            | .          | PB.3       | PB.3.1          | bc1001          | PacBio     | exon        | 1393395   | 1393460   | .          | -            | .          | ...   |\n",
       "| ...          | ...        | ...        | ...       | ...       | ...        | ...          | ...        | ...        | ...             | ...             | ...        | ...         | ...       | ...       | ...        | ...          | ...        | ...   |\n",
       "| chr1         | PacBio     | exon       | 1390767   | 1394777   | .          | +            | .          | PB.3       | PB.3.1          | bc1001          | PacBio     | transcript  | 1390334   | 1399339   | .          | -            | .          | ...   |\n",
       "| chr1         | PacBio     | exon       | 1390767   | 1394777   | .          | +            | .          | PB.3       | PB.3.1          | bc1001          | PacBio     | exon        | 1390765   | 1392803   | .          | -            | .          | ...   |\n",
       "| chr1         | PacBio     | exon       | 1390767   | 1394777   | .          | +            | .          | PB.3       | PB.3.1          | bc1001          | PacBio     | exon        | 1390765   | 1392803   | .          | -            | .          | ...   |\n",
       "| chr1         | PacBio     | exon       | 1390767   | 1394777   | .          | +            | .          | PB.3       | PB.3.1          | bc1001          | PacBio     | transcript  | 1386710   | 1399335   | .          | -            | .          | ...   |\n",
       "+--------------+------------+------------+-----------+-----------+------------+--------------+------------+------------+-----------------+-----------------+------------+-------------+-----------+-----------+------------+--------------+------------+-------+\n",
       "Stranded PyRanges object has 108 rows and 22 columns from 1 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "4 hidden columns: gene_id_b, transcript_id_b, sample_source_b, Cluster"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bc1001_1002.cluster().subset(lambda x: x.Cluster == 1).subset(lambda y: y.)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "8d28ca0c-9b3d-4665-ab28-ef8ee9564feb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "5311945"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(bc1001_1002)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "83546034-48b7-4861-9a0d-538d82463cd7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "208502"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(bc1001_annot)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "33373999-0841-476e-ad58-24752670408c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "172893"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(bc1002_annot)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d04a08ca-e119-4dc1-b2e4-b5552083faef",
   "metadata": {},
   "source": [
    "Not sure why there are ~20-50 times as many rows in the merged set as there are in the two parental non-merged objects.\n",
    "\n",
    "Another way of doing this would be to concatinate the files and then import and cluster."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "92f257d4-f742-4d0c-aca1-2e7b8a9f6cd2",
   "metadata": {},
   "outputs": [],
   "source": [
    "bc1001_annot.to_gff3(path=\"alt_bc1001.gff3\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "6f618262-d025-4e2c-8970-61dff1d0e766",
   "metadata": {},
   "outputs": [],
   "source": [
    "bc1002_annot.to_gff3(path=\"alt_bc1002.gff3\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "22954092-7492-49cd-9d2a-de685b90b795",
   "metadata": {},
   "outputs": [],
   "source": [
    "!cat alt_bc1001.gff3 alt_bc1002.gff3 > bc1001-bc1002-merged.gff3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "3b605be2-95be-4fe5-9366-3e6e57d9e039",
   "metadata": {},
   "outputs": [],
   "source": [
    "bc1001_sample_name = pd.Series([\"bc1001\" for _ in range(len(bc1001_annot))], name=\"Source\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "6917562b-9b9d-4388-9502-7dd696d147df",
   "metadata": {},
   "outputs": [],
   "source": [
    "bc1001_annot.drop(\"sample_source\").insert(bc1001_sample_name).to_gff3(path=\"alt_bc1001.gff3\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "0421a643-f9ee-4567-a8f1-99de12cc2c9e",
   "metadata": {},
   "outputs": [],
   "source": [
    "bc1002_sample_name = pd.Series([\"bc1002\" for _ in range(len(bc1002_annot))], name=\"Source\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "7397dfe8-c24a-4344-a1d5-55f28d64d12b",
   "metadata": {},
   "outputs": [],
   "source": [
    "bc1002_annot.drop(\"sample_source\").insert(bc1002_sample_name).to_gff3(path=\"alt_bc1002.gff3\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "bf42fa1a-a9fc-4c73-8447-5f353a3d26aa",
   "metadata": {},
   "outputs": [],
   "source": [
    "!cat alt_bc1001.gff3 alt_bc1002.gff3 > bc1001-bc1002-merged.gff3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "b3e292df-a173-4244-ad79-469ffb4a0ae6",
   "metadata": {},
   "outputs": [],
   "source": [
    "merged = pr.read_gtf(\"/home/milo/workspace/cdna_brownie/notebooks/test_data/all-merged.gff\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a7d0e81b-2453-4c2e-b1c7-73e094151424",
   "metadata": {},
   "source": [
    "Some of the clusters appear to have **very** cover very wide intervals, to the point where I'm uyn "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "98d31905-3baa-4430-b955-c3a8d361c5dd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>Cluster</th>\n",
       "      <th>Count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>bc1004</td>\n",
       "      <td>transcript</td>\n",
       "      <td>2556370</td>\n",
       "      <td>2563747</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PB.13</td>\n",
       "      <td>PB.13.2</td>\n",
       "      <td>2</td>\n",
       "      <td>44</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>bc1004</td>\n",
       "      <td>transcript</td>\n",
       "      <td>2555793</td>\n",
       "      <td>4508611</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PB.13</td>\n",
       "      <td>PB.13.1</td>\n",
       "      <td>2</td>\n",
       "      <td>44</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+------------+--------------+-----------+-----------+------------+--------------+------------+------------+-----------------+-----------+-----------+\n",
       "| Chromosome   | Source     | Feature      |     Start |       End | Score      | Strand       | Frame      | gene_id    | transcript_id   |   Cluster |     Count |\n",
       "| (category)   | (object)   | (category)   |   (int32) |   (int32) | (object)   | (category)   | (object)   | (object)   | (object)        |   (int32) |   (int64) |\n",
       "|--------------+------------+--------------+-----------+-----------+------------+--------------+------------+------------+-----------------+-----------+-----------|\n",
       "| chr1         | bc1004     | transcript   |   2556370 |   2563747 | .          | +            | .          | PB.13      | PB.13.2         |         2 |        44 |\n",
       "| chr1         | bc1004     | transcript   |   2555793 |   4508611 | .          | +            | .          | PB.13      | PB.13.1         |         2 |        44 |\n",
       "+--------------+------------+--------------+-----------+-----------+------------+--------------+------------+------------+-----------------+-----------+-----------+\n",
       "Stranded PyRanges object has 2 rows and 12 columns from 1 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand."
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged.subset(lambda x: x.Source == \"bc1004\").subset(lambda y: y.gene_id == \"PB.13\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "5e7bca41-b002-4ebe-a2e0-eb2784c6caa9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>Cluster</th>\n",
       "      <th>Count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr1</td>\n",
       "      <td>bc1019</td>\n",
       "      <td>transcript</td>\n",
       "      <td>3857266</td>\n",
       "      <td>3885426</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PB.7</td>\n",
       "      <td>PB.7.1</td>\n",
       "      <td>2</td>\n",
       "      <td>44</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+------------+--------------+-----------+-----------+------------+--------------+------------+------------+-----------------+-----------+-----------+\n",
       "| Chromosome   | Source     | Feature      |     Start |       End | Score      | Strand       | Frame      | gene_id    | transcript_id   |   Cluster |     Count |\n",
       "| (category)   | (object)   | (category)   |   (int32) |   (int32) | (object)   | (category)   | (object)   | (object)   | (object)        |   (int32) |   (int64) |\n",
       "|--------------+------------+--------------+-----------+-----------+------------+--------------+------------+------------+-----------------+-----------+-----------|\n",
       "| chr1         | bc1019     | transcript   |   3857266 |   3885426 | .          | +            | .          | PB.7       | PB.7.1          |         2 |        44 |\n",
       "+--------------+------------+--------------+-----------+-----------+------------+--------------+------------+------------+-----------------+-----------+-----------+\n",
       "Stranded PyRanges object has 1 rows and 12 columns from 1 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand."
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged.subset(lambda x: x.Source == \"bc1019\").subset(lambda y: y.gene_id == \"PB.7\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4f0b9321-2ea4-40c3-8004-6b0c167c223a",
   "metadata": {},
   "source": [
    "This makes a bit more sense now: bc1004 PB.13.2 is a subset of PB.13.1 and bc1019 PB.7.1 is a subset of bc1004 PB.13.2"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "993fd416-b0ba-4034-a572-3e2c6cc70d31",
   "metadata": {},
   "source": [
    "So, to summarize what we need to do:\n",
    "\n",
    "1) In each GFF, change the \"Source\" field to the sample name.\n",
    "2) Concatinate all GFFs.\n",
    "3) Cluster the transcripts/exons in the merged PyRanges object.\n",
    "4) Create a new PBID based on the cluster number and use this to convert the existing gene_id and transcript_ids.\n",
    "5) For each distinct \"Source\", we need to create a translation table mapping the previous gene_is/transcript_id and the new cluster-based ones.\n",
    "6) Convert the ids in the other associated files (i.e. the abundance and group files) for each sample"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "4e0acf78-173e-4746-82f5-c056cc449713",
   "metadata": {},
   "outputs": [],
   "source": [
    "prdf = merged.as_df()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "836547c1-57a7-41a0-a7e3-6d57fd4ab446",
   "metadata": {},
   "outputs": [],
   "source": [
    "pr_dict = prdf.to_dict()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "72b92b11-733b-4344-ab11-2d3782a4f2c9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>Cluster</th>\n",
       "      <th>Count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>bc1004</td>\n",
       "      <td>transcript</td>\n",
       "      <td>833846</td>\n",
       "      <td>838936</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PB.3</td>\n",
       "      <td>PB.3.1</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>bc1003</td>\n",
       "      <td>transcript</td>\n",
       "      <td>827658</td>\n",
       "      <td>857173</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PB.2</td>\n",
       "      <td>PB.2.1</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>bc1002</td>\n",
       "      <td>transcript</td>\n",
       "      <td>854801</td>\n",
       "      <td>859446</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PB.2</td>\n",
       "      <td>PB.2.1</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>bc1005</td>\n",
       "      <td>transcript</td>\n",
       "      <td>834239</td>\n",
       "      <td>838441</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PB.1</td>\n",
       "      <td>PB.1.1</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>bc1004</td>\n",
       "      <td>transcript</td>\n",
       "      <td>2556370</td>\n",
       "      <td>2563747</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PB.13</td>\n",
       "      <td>PB.13.2</td>\n",
       "      <td>2</td>\n",
       "      <td>44</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>175277</th>\n",
       "      <td>chrY</td>\n",
       "      <td>bc1012</td>\n",
       "      <td>transcript</td>\n",
       "      <td>19684562</td>\n",
       "      <td>19744726</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PB.3466</td>\n",
       "      <td>PB.3466.1</td>\n",
       "      <td>3836</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>175278</th>\n",
       "      <td>chrY</td>\n",
       "      <td>bc1012</td>\n",
       "      <td>transcript</td>\n",
       "      <td>19684566</td>\n",
       "      <td>19744723</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PB.3466</td>\n",
       "      <td>PB.3466.2</td>\n",
       "      <td>3836</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>175279</th>\n",
       "      <td>chrY</td>\n",
       "      <td>bc1012</td>\n",
       "      <td>transcript</td>\n",
       "      <td>19705423</td>\n",
       "      <td>19744726</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PB.3466</td>\n",
       "      <td>PB.3466.3</td>\n",
       "      <td>3836</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>175280</th>\n",
       "      <td>chrY</td>\n",
       "      <td>bc1012</td>\n",
       "      <td>transcript</td>\n",
       "      <td>19705425</td>\n",
       "      <td>19744723</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PB.3466</td>\n",
       "      <td>PB.3466.4</td>\n",
       "      <td>3836</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>175281</th>\n",
       "      <td>chrY</td>\n",
       "      <td>bc1012</td>\n",
       "      <td>transcript</td>\n",
       "      <td>19705601</td>\n",
       "      <td>19744714</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PB.3466</td>\n",
       "      <td>PB.3466.5</td>\n",
       "      <td>3836</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>175282 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       Chromosome  Source     Feature     Start       End Score Strand Frame  \\\n",
       "0            chr1  bc1004  transcript    833846    838936     .      +     .   \n",
       "1            chr1  bc1003  transcript    827658    857173     .      +     .   \n",
       "2            chr1  bc1002  transcript    854801    859446     .      +     .   \n",
       "3            chr1  bc1005  transcript    834239    838441     .      +     .   \n",
       "4            chr1  bc1004  transcript   2556370   2563747     .      +     .   \n",
       "...           ...     ...         ...       ...       ...   ...    ...   ...   \n",
       "175277       chrY  bc1012  transcript  19684562  19744726     .      -     .   \n",
       "175278       chrY  bc1012  transcript  19684566  19744723     .      -     .   \n",
       "175279       chrY  bc1012  transcript  19705423  19744726     .      -     .   \n",
       "175280       chrY  bc1012  transcript  19705425  19744723     .      -     .   \n",
       "175281       chrY  bc1012  transcript  19705601  19744714     .      -     .   \n",
       "\n",
       "        gene_id transcript_id  Cluster  Count  \n",
       "0          PB.3        PB.3.1        1      4  \n",
       "1          PB.2        PB.2.1        1      4  \n",
       "2          PB.2        PB.2.1        1      4  \n",
       "3          PB.1        PB.1.1        1      4  \n",
       "4         PB.13       PB.13.2        2     44  \n",
       "...         ...           ...      ...    ...  \n",
       "175277  PB.3466     PB.3466.1     3836      5  \n",
       "175278  PB.3466     PB.3466.2     3836      5  \n",
       "175279  PB.3466     PB.3466.3     3836      5  \n",
       "175280  PB.3466     PB.3466.4     3836      5  \n",
       "175281  PB.3466     PB.3466.5     3836      5  \n",
       "\n",
       "[175282 rows x 12 columns]"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.DataFrame(data = pr_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "f89bcf48-de1a-4671-8fb6-f3723eee5146",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>Cluster</th>\n",
       "      <th>Count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>bc1004</td>\n",
       "      <td>transcript</td>\n",
       "      <td>833846</td>\n",
       "      <td>838936</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PB.3</td>\n",
       "      <td>PB.3.1</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>bc1003</td>\n",
       "      <td>transcript</td>\n",
       "      <td>827658</td>\n",
       "      <td>857173</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PB.2</td>\n",
       "      <td>PB.2.1</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>bc1002</td>\n",
       "      <td>transcript</td>\n",
       "      <td>854801</td>\n",
       "      <td>859446</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PB.2</td>\n",
       "      <td>PB.2.1</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>bc1005</td>\n",
       "      <td>transcript</td>\n",
       "      <td>834239</td>\n",
       "      <td>838441</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PB.1</td>\n",
       "      <td>PB.1.1</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>bc1004</td>\n",
       "      <td>transcript</td>\n",
       "      <td>2556370</td>\n",
       "      <td>2563747</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PB.13</td>\n",
       "      <td>PB.13.2</td>\n",
       "      <td>2</td>\n",
       "      <td>44</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>175277</th>\n",
       "      <td>chrY</td>\n",
       "      <td>bc1012</td>\n",
       "      <td>transcript</td>\n",
       "      <td>19684562</td>\n",
       "      <td>19744726</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PB.3466</td>\n",
       "      <td>PB.3466.1</td>\n",
       "      <td>3836</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>175278</th>\n",
       "      <td>chrY</td>\n",
       "      <td>bc1012</td>\n",
       "      <td>transcript</td>\n",
       "      <td>19684566</td>\n",
       "      <td>19744723</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PB.3466</td>\n",
       "      <td>PB.3466.2</td>\n",
       "      <td>3836</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>175279</th>\n",
       "      <td>chrY</td>\n",
       "      <td>bc1012</td>\n",
       "      <td>transcript</td>\n",
       "      <td>19705423</td>\n",
       "      <td>19744726</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PB.3466</td>\n",
       "      <td>PB.3466.3</td>\n",
       "      <td>3836</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>175280</th>\n",
       "      <td>chrY</td>\n",
       "      <td>bc1012</td>\n",
       "      <td>transcript</td>\n",
       "      <td>19705425</td>\n",
       "      <td>19744723</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PB.3466</td>\n",
       "      <td>PB.3466.4</td>\n",
       "      <td>3836</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>175281</th>\n",
       "      <td>chrY</td>\n",
       "      <td>bc1012</td>\n",
       "      <td>transcript</td>\n",
       "      <td>19705601</td>\n",
       "      <td>19744714</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>PB.3466</td>\n",
       "      <td>PB.3466.5</td>\n",
       "      <td>3836</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>175282 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+------------+------------+-----------+-----------+------------+--------------+------------+------------+-----------------+-----------+-----------+\n",
       "| Chromosome   | Source     | Feature    | Start     | End       | Score      | Strand       | Frame      | gene_id    | transcript_id   | Cluster   | Count     |\n",
       "| (category)   | (object)   | (object)   | (int32)   | (int32)   | (object)   | (category)   | (object)   | (object)   | (object)        | (int64)   | (int64)   |\n",
       "|--------------+------------+------------+-----------+-----------+------------+--------------+------------+------------+-----------------+-----------+-----------|\n",
       "| chr1         | bc1004     | transcript | 833846    | 838936    | .          | +            | .          | PB.3       | PB.3.1          | 1         | 4         |\n",
       "| chr1         | bc1003     | transcript | 827658    | 857173    | .          | +            | .          | PB.2       | PB.2.1          | 1         | 4         |\n",
       "| chr1         | bc1002     | transcript | 854801    | 859446    | .          | +            | .          | PB.2       | PB.2.1          | 1         | 4         |\n",
       "| chr1         | bc1005     | transcript | 834239    | 838441    | .          | +            | .          | PB.1       | PB.1.1          | 1         | 4         |\n",
       "| ...          | ...        | ...        | ...       | ...       | ...        | ...          | ...        | ...        | ...             | ...       | ...       |\n",
       "| chrY         | bc1012     | transcript | 19684566  | 19744723  | .          | -            | .          | PB.3466    | PB.3466.2       | 3836      | 5         |\n",
       "| chrY         | bc1012     | transcript | 19705423  | 19744726  | .          | -            | .          | PB.3466    | PB.3466.3       | 3836      | 5         |\n",
       "| chrY         | bc1012     | transcript | 19705425  | 19744723  | .          | -            | .          | PB.3466    | PB.3466.4       | 3836      | 5         |\n",
       "| chrY         | bc1012     | transcript | 19705601  | 19744714  | .          | -            | .          | PB.3466    | PB.3466.5       | 3836      | 5         |\n",
       "+--------------+------------+------------+-----------+-----------+------------+--------------+------------+------------+-----------------+-----------+-----------+\n",
       "Stranded PyRanges object has 175,282 rows and 12 columns from 35 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand."
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pr.from_dict(pr_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "2358229c-1ec0-4bd2-85a5-94a88b08dea2",
   "metadata": {},
   "outputs": [],
   "source": [
    "merged2 = merged.copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "62c09732-78bb-4316-878e-a52509cb5712",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>Cluster</th>\n",
       "      <th>Count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>stuff</td>\n",
       "      <td>transcript</td>\n",
       "      <td>833846</td>\n",
       "      <td>838936</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PB.3</td>\n",
       "      <td>PB.3.1</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>stuff</td>\n",
       "      <td>transcript</td>\n",
       "      <td>827658</td>\n",
       "      <td>857173</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PB.2</td>\n",
       "      <td>PB.2.1</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>stuff</td>\n",
       "      <td>transcript</td>\n",
       "      <td>854801</td>\n",
       "      <td>859446</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PB.2</td>\n",
       "      <td>PB.2.1</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>stuff</td>\n",
       "      <td>transcript</td>\n",
       "      <td>834239</td>\n",
       "      <td>838441</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PB.1</td>\n",
       "      <td>PB.1.1</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>stuff</td>\n",
       "      <td>transcript</td>\n",
       "      <td>2556370</td>\n",
       "      <td>2563747</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PB.13</td>\n",
       "      <td>PB.13.2</td>\n",
       "      <td>2</td>\n",
       "      <td>44</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr1</td>\n",
       "      <td>stuff</td>\n",
       "      <td>transcript</td>\n",
       "      <td>3857266</td>\n",
       "      <td>3885426</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PB.7</td>\n",
       "      <td>PB.7.1</td>\n",
       "      <td>2</td>\n",
       "      <td>44</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr1</td>\n",
       "      <td>stuff</td>\n",
       "      <td>transcript</td>\n",
       "      <td>3857266</td>\n",
       "      <td>3885424</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PB.12</td>\n",
       "      <td>PB.12.1</td>\n",
       "      <td>2</td>\n",
       "      <td>44</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chr1</td>\n",
       "      <td>stuff</td>\n",
       "      <td>transcript</td>\n",
       "      <td>2556213</td>\n",
       "      <td>2563742</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>PB.9</td>\n",
       "      <td>PB.9.1</td>\n",
       "      <td>2</td>\n",
       "      <td>44</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+------------+--------------+-----------+-----------+------------+--------------+------------+------------+-----------------+-----------+-----------+\n",
       "| Chromosome   | Source     | Feature      |     Start |       End | Score      | Strand       | Frame      | gene_id    | transcript_id   |   Cluster |     Count |\n",
       "| (category)   | (object)   | (category)   |   (int32) |   (int32) | (object)   | (category)   | (object)   | (object)   | (object)        |   (int32) |   (int64) |\n",
       "|--------------+------------+--------------+-----------+-----------+------------+--------------+------------+------------+-----------------+-----------+-----------|\n",
       "| chr1         | stuff      | transcript   |    833846 |    838936 | .          | +            | .          | PB.3       | PB.3.1          |         1 |         4 |\n",
       "| chr1         | stuff      | transcript   |    827658 |    857173 | .          | +            | .          | PB.2       | PB.2.1          |         1 |         4 |\n",
       "| chr1         | stuff      | transcript   |    854801 |    859446 | .          | +            | .          | PB.2       | PB.2.1          |         1 |         4 |\n",
       "| chr1         | stuff      | transcript   |    834239 |    838441 | .          | +            | .          | PB.1       | PB.1.1          |         1 |         4 |\n",
       "| chr1         | stuff      | transcript   |   2556370 |   2563747 | .          | +            | .          | PB.13      | PB.13.2         |         2 |        44 |\n",
       "| chr1         | stuff      | transcript   |   3857266 |   3885426 | .          | +            | .          | PB.7       | PB.7.1          |         2 |        44 |\n",
       "| chr1         | stuff      | transcript   |   3857266 |   3885424 | .          | +            | .          | PB.12      | PB.12.1         |         2 |        44 |\n",
       "| chr1         | stuff      | transcript   |   2556213 |   2563742 | .          | +            | .          | PB.9       | PB.9.1          |         2 |        44 |\n",
       "+--------------+------------+--------------+-----------+-----------+------------+--------------+------------+------------+-----------------+-----------+-----------+\n",
       "Stranded PyRanges object has 8 rows and 12 columns from 1 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand."
      ]
     },
     "execution_count": 62,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged2 = merged2.insert(pd.Series([\"stuff\" for _ in range(len(merged2.as_df()))], name=\"Source\"))\n",
    "merged2.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "08477501-7952-443e-ae92-635ea8321833",
   "metadata": {},
   "outputs": [],
   "source": [
    "merged2.insert()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9cca9e7e-be1f-419d-9578-36361f729167",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
